scripts/affiliate-server

```diff
@@ -0,0 +1,222 @@
+        exceededed (i.e. wait seconds passed)
+        """
+
+        if not web.amazon_api:
+            return json.dumps({"error": "not_configured"})
+
+        isbn10, isbn13 = self.unpack_isbn(isbn)
+        if not (isbn10 and isbn13):
+            return json.dumps({
+                "error": "rejected_isbn",
+                "isbn10": isbn10,
+                "isbn13": isbn13
+            })
+
+        # Cache lookup by isbn13. If there's a miss, add to queue
+        if not cache.memcache_cache.get('amazon_product_%s' % isbn13):
+            # Misses (which will get submitted to amz as ASIN) are isbn10
+            # We should track when (queue % 20) occurs
+            web.amazon_queue.add(isbn10)
+
+        threshold_passed = len(web.amazon_queue) > threshold
+        timer_passed = time.time() + wait > web.amazon_timer
+        if len(web.amazon_queue) and (threshold_passed or timer_passed):
+            # amz has a limit of 10 IDs but the queue may have more than 10 ids
+            # so we should only pop 10 items off the queue and keep the rest on
+            queue = list(web.amazon_queue)
+            web.amazon_queue = set(queue[10:])
+            try:
+                products = web.amazon_api.get_products(queue[:10], serialize=True)
+                for product in products:
+                    # Make sure we have an isbn13 for cache key
+                    cache_key = (
+                        product.get('isbn_13') and product.get('isbn_13')[0] or 
+                        isbn_10_to_isbn_13(product.get('isbn_10')[0])
+                    )
+                    cache.memcache_cache.set(
+                        'amazon_product_%s' % cache_key, product, expires=DAY_SECS)
+                # Reset the amazon timer
+                web.amazon_timer = time.time()
+            except Exception:
+                logger.exception("amazon_api.get_product({}, serialize=True)".format(queue))                
+                web.amazon_queue.update(queue)  # Careful, request may enter fail-loop from bad IDs
+
+        return json.dumps({
+            "status": "success",
+            "hit": cache.memcache_cache.get('amazon_product_%s' % isbn13)
+        } if cache.memcache_cache.get('amazon_product_%s' % isbn13) else {
+            "status": "submitted",
+            "queue": len(web.amazon_queue),
+            "threshold": threshold,
+            "timer": web.amazon_timer
+        })
+
+def load_config(configfile):
+    with open(configfile) as in_file:
+        d = yaml.safe_load(in_file)
+    for k, v in d.items():
+        setattr(web.ol_config, k, v)
+
+    if 'fastcgi' in d:
+        web.config.fastcgi = d['fastcgi']
+
+    web.amazon_api = None
+    args = [
+        web.ol_config.amazon_api.get('key'),
+        web.ol_config.amazon_api.get('secret'),
+        web.ol_config.amazon_api.get('id')
+    ]
+    if all(args):
+        web.amazon_api = AmazonAPI(*args, throttling=0.9)
+        web.amazon_queue = set()
```

Comment:
My sense is that we should _really_ use a queue... https://docs.python.org/3/library/queue.html

Python sets are guaranteed to be unordered.  Python queues are FIFO which is I think what we want.

At the top of the file: `from queue import Queue`
```suggestion
        web.amazon_queue = Queue()
```
If the value is in Memcached then return it
else `web.amazon_queue.put_nowait(value)` and return a promise

... acquision logic ...
batch = set()  # removes any dups...
while not amazon_queue.empty() and len(batch) < threshold:
    batch.add(amazon_queue.get())
if batch:
    amazon_lookup(batch)